File: Lecture26/poisson.py
import math
import time
import cpt

class Poisson :
    def __init__ (self, L=50):
        self.L = L                   # number of interior points in x and y
        self.omega = 1.88177         # over-relaxation parameter for L = 50
        self.N = L + 2               # interior plus two boundary points
        N=self.N
        self.V = cpt.Matrix(N, N)    # potential to be found
        self.rho = cpt.Matrix(N, N)  # given charge density
        self.VNew = cpt.Matrix(N, N) # new potential after each step
        self.h = 1.0 / (L + 1)       # lattice spacing assuming size in x and y = 1
        self.q = 10.0                # point charge
        i = N / 2                    # center of lattice
        self.rho[i][i] = self.q / self.h**2    # charge density


    def Jacobi(self) :               # Jacobi algorithm for a single iterative step
        VNew = self.VNew
        V    = self.V         #avoid lots of typing
        rho  = self.rho
        h    = self.h
        for i in range(1, self.L+1):
            for j in range(1, self.L+1):
                VNew[i][j] = 0.25 * (V[i-1][j] + V[i+1][j] +
                                     V[i][j-1] + V[i][j+1] +
                                     h**2 * rho[i][j])

    def GaussSeidel(self):          # Gauss-Seidel algorithm for one iterative step
        L = self.L
        VNew = self.VNew
        V    = self.V         #avoid lots of typing
        rho  = self.rho
        h    = self.h        

        # copy V to VNew
        for i in range(1, self.L+1):
            for j in range(1, self.L+1):
                VNew[i][j] =  V[i][j]

        # perform Gauss-Seidel update
        for i in range(1, self.L):
            for j in range(1, self.L+1):
                VNew[i][j] = 0.25 * (VNew[i-1][j] + VNew[i+1][j] +
                                     VNew[i][j-1] + VNew[i][j+1] +
                                     h**2 * rho[i][j])


    def SuccessiveOverRelaxation(self):
        L = self.L
        VNew = self.VNew
        V    = self.V         #avoid lots of typing
        rho  = self.rho
        h    = self.h
        omega= self.omega

        # update even sites in red-black scheme
        for i in range(1, self.L+1):
            for j in range(1, self.L+1):
                if (i + j) % 2 == 0:
                    VNew[i][j] = (1 - omega) * V[i][j] + omega / 4 * (
                                 V[i-1][j] + V[i+1][j] + V[i][j-1] +
                                 V[i][j+1] + h**2 * rho[i][j] )

        # update odd sites in red-black scheme
        for i in range(1, self.L+1):
            for j in range(1, self.L+1):
                if (i + j) % 2 != 0:
                    VNew[i][j] = (1 - omega) * V[i][j] + omega / 4 * (
                                 VNew[i-1][j] + VNew[i+1][j] + VNew[i][j-1] +
                                 VNew[i][j+1] + h**2 * rho[i][j] )

    def relativeError(self):
        L = self.L
        VNew = self.VNew
        V    = self.V         #avoid lots of typing
        rho  = self.rho
        h    = self.h
        omega= self.omega
        
        error = 0
        n = 0
        for i in range(1, self.L+1):
            for j in range(1, self.L+1):
                if VNew[i][j] != 0 and VNew[i][j] != V[i][j]:
                    error += abs(1 - V[i][j] / VNew[i][j])
                    n += 1
        if n != 0:
            error /= n
        return error





print " Iterative solution of Poisson's equation"
print " ----------------------------------------"
L = int(input(" Enter number of interior points in x or y: "))
poisson = Poisson(L=L)
accuracy = float(input(" Enter desired accuracy in solution: "))
choice = int( input(" Enter choice of algorithm, Jacobi (0), Gauss-Seidel (1) or Successive overrelaxation (2): ") )

start_time = time.clock()

steps = 0

while True:
    if choice == 0 : 
        poisson.Jacobi()
    elif choice == 1 : 
        poisson.GaussSeidel()
    else : 
        poisson.SuccessiveOverRelaxation()
    if poisson.relativeError() < accuracy:
        break
    for i in range(1, poisson.L+1):
        for j in range(1, poisson.L+1):
            poisson.V[i][j] = poisson.VNew[i][j]
    steps += 1

print " Number of steps =", steps
print " CPU time =", time.clock() - start_time, "sec"


# Convert x, y, V(x,y) to a surface plot
from mpl_toolkits.mplot3d import Axes3D
from matplotlib import cm
from matplotlib.ticker import LinearLocator, FormatStrFormatter
from matplotlib.mlab import griddata
import matplotlib.pyplot as plt
import numpy as np
# Define the axes
x = np.arange(0, poisson.h*(poisson.L+2), poisson.h)
y = np.arange(0, poisson.h*(poisson.L+2), poisson.h)
# Get the grid
X, Y = np.meshgrid(x, y)
# Set Z to the poisson V[i][j]
Z = np.array( poisson.V )

fig = plt.figure()
ax = fig.gca(projection='3d')
scat = ax.plot_surface( X, Y, Z, rstride=1, cstride=1, cmap=cm.coolwarm,
                        linewidth=0, antialiased=False )

plt.show()

File: Lecture26/poisson.cpp
#include "cptstd.hpp"
#include "matrix.hpp"
using namespace cpt;

class Poisson {
public : 
  Poisson( int iL = 50 )  : 
    L(iL)
  {
    int N = L + 2;
    V = rho = V_new = Matrix<double,2>(N, N);


    h = 1 / double(L + 1);      // assume physical size in x and y = 1
    double q = 10;              // point charge
    int i = N / 2;              // center of lattice
    rho[i][i] = q / (h * h);    // charge density
    steps = 0;
  }

  void Jacobi() {
    // Jacobi algorithm for a single iterative step
    for (int i = 1; i <= L; i++)
      for (int j = 1; j <= L; j++)
	V_new[i][j] = 0.25 * (V[i - 1][j] + V[i + 1][j] +
			      V[i][j - 1] + V[i][j + 1] +
			      h * h * rho[i][j]);
  }
  double relative_error()
  {
    double error = 0;           // average relative error per lattice point
    int n = 0;                  // number of non-zero differences

    for (int i = 1; i <= L; i++)
      for (int j = 1; j <= L; j++) {
	if (V_new[i][j] != 0)
	  if (V_new[i][j] != V[i][j]) {
	    error += abs(1 - V[i][j] / V_new[i][j]);
	    ++n;
	  }
      }
    if (n != 0)
      error /= n;
    return error;
  }

  void Gauss_Seidel()
  {
    // copy V to V_new
    V_new = V;

    // Gauss-Seidel update in place
    for (int i = 1; i <= L; i++)
      for (int j = 1; j <= L; j++)
	V_new[i][j] = 0.25 * (V_new[i - 1][j] + V_new[i + 1][j] +
			      V_new[i][j - 1] + V_new[i][j + 1] +
			      h * h * rho[i][j]);
  }

  void successive_over_relaxation()   // using red-black checkerboard updating
  {
    // update even sites first
    for (int i = 1; i <= L; i++)
      for (int j = 1; j <= L; j++)
	if ((i + j) % 2 == 0)
	  V_new[i][j] = (1 - omega) * V[i][j] + omega / 4 *
	    (V[i - 1][j] + V[i + 1][j] +
	     V[i][j - 1] + V[i][j + 1] +
	     h * h * rho[i][j]);

    // update odd sites using updated even sites
    for (int i = 1; i <= L; i++)
      for (int j = 1; j <= L; j++)
	if ((i + j) % 2 != 0)
	  V_new[i][j] = (1 - omega) * V[i][j] + omega / 4 *
	    (V_new[i - 1][j] + V_new[i + 1][j] +
	     V_new[i][j - 1] + V_new[i][j + 1] +
	     h * h * rho[i][j]);
  }

  template< typename T> 
  void iterate( T const & method,double accuracy)
  {
    clock_t t0 = clock();

    while (true) {
      (this->*method)();
      ++steps;
      double error = relative_error();
      if (error < accuracy)
	break;
      swap(V, V_new);         // use <algorithm> std::swap
    }
    cout << " Number of steps = " << steps << endl;

    clock_t t1 = clock();
    cout << " CPU time = " << double(t1 - t0) / CLOCKS_PER_SEC
         << " sec" << endl;
  }

  void set_omega ( double i ) { omega = i; }

  double get_h() const { return h; }
  int    get_L() const { return L; }
  double get_V( int i, int j) const { return V[i][j]; }

protected : 
  int L ;                         // number of interior points in x and y
  Matrix<double,2> V,             // potential to be found
    rho,                          // given charge density
    V_new;                        // new potential after each step

  double h;                       // lattice spacing
  int steps;                      // number of iteration steps
  double accuracy;                // desired accuracy in solution
  double omega;                   // overrelaxation parameter


};

int main() {

  int L;
    cout << " Iterative solution of Poisson's equation\n"
         << " ----------------------------------------\n";
    cout << " Enter number of interior points in x or y: ";
    cin >> L;

    Poisson p(L);

    double accuracy; 
    cout << " Enter desired accuracy in solution: ";
    cin >> accuracy;
    cout << " Enter 0 for Jacobi, 1 for Gauss Seidel, 2 for SOR: ";
    int choice;
    cin >> choice;

    void (Poisson::* ptfptr) (void);
    switch (choice) {
    case 0:
      ptfptr = &Poisson::Jacobi;
      p.iterate(ptfptr, accuracy);

      break;
    case 1:
      ptfptr = &Poisson::Gauss_Seidel;
      p.iterate(ptfptr, accuracy);

      break;
    case 2: default : 
      ptfptr = &Poisson::successive_over_relaxation;
      double omega = 2 / (1 + 4 * atan(1.0) / double(L));
      p.set_omega( omega ) ;
      p.iterate(ptfptr, accuracy);

      break;
    }


    // write potential to file
    cout << " Potential in file poisson.data" << endl;
    ofstream date_file("poisson.data");
    for (int i = 0; i < p.get_L() + 2; i++) {
      double x = i * p.get_h();
      for (int j = 0; j < p.get_L() + 2; j++) {
	double y = j * p.get_h();
	char buff[1000];
	sprintf(buff, "%12.6f %12.6f %12.6f", x, y, p.get_V(i,j) );
	date_file << buff << endl;
      }
    }
    date_file.close();
}
